simple_history/manager.py

```diff
@@ -89,14 +90,38 @@ def _as_of_set(self, date):
         model = type(self.model().instance)  # a bit of a hack to get the model
         pk_attr = model._meta.pk.name
         queryset = self.get_queryset().filter(history_date__lte=date)
-        for original_pk in set(queryset.order_by().values_list(pk_attr, flat=True)):
-            changes = queryset.filter(**{pk_attr: original_pk})
-            last_change = changes.latest("history_date")
-            if changes.filter(
-                history_date=last_change.history_date, history_type="-"
-            ).exists():
-                continue
-            yield last_change.instance
+        # If using MySQL, need to get a list of IDs in memory and then use them for the
+        # second query.
+        # Does mean two loops through the DB to get the full set, but still a speed
+        # improvement.
+        backend = connection.vendor
+        if backend == "mysql":
+            history_ids = {}
+            for item in queryset.order_by("-history_date", "-pk"):
+                if getattr(item, pk_attr) not in history_ids:
+                    history_ids[getattr(item, pk_attr)] = item.pk
```

Comment:
not sure I understand this part?

Reply:
The original method pulled a set of `pk_attr`, then for each one in the set did a filter & latest to get the most recent historical item per `pk_attr`.

This way instead only loops through once - ordering the history set by `-history_date,-pk` (for same date conflicts), and taking the first historical entry found for each `pk_attr`, putting it into the dict to use in the next part.

Hope that explained it well.